### ISSUE TITLE
Scroll Update Access Container Discontinues

### ISSUE THREAD

**Issue by arz-GW** [moby/moby#39459]
Description
when i use "docker stack deploy" to run a container on a node, and then Rolling update service, Continuous access to containers found transient connection failure.
I set up rolling updates and health checks, and the problem remains.

Steps to reproduce the issue:
     1.the yaml is like this:

```yaml
version: "3"
services:
  test-healthcheck:
    image: test-healthcheck-old:1
    ports:
      -  "2345:80"
    networks:
      - arz_net
    deploy:
      replicas: 2 
      update_config:
        parallelism: 1
        delay: 1m
      restart_policy:
        condition: on-failure
    healthcheck:
      test: ["CMD", "curl", "-f", "http://127.0.0.1 || exit 1"]
      interval: 10s
      timeout: 5s
      retries: 3
networks:
  arz_net:
    external: true
```

2. Rolling update service

```
docker stack deploy -c test-healthcheck.yaml  test-healthcheck
```

3. Uninterrupted access to containers

```
 while true; do curl 127.0.0.1:2345 ;done

11111
11111
11111
11111
curl: (56) Recv failure: Connection reset by peer
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
11111
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
.....
new
new
```

Describe the results you expected:

i want to know the the cause of the problem and the ways to solve the problem.
I appreciate this very much

Additional information you deem important (e.g. issue happens only occasionally):

Output of docker version:

```
Client:
 Version:      17.06.2-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 19:59:06 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   cec0b72
 Built:        Tue Sep  5 20:00:25 2017
 OS/Arch:      linux/amd64
 Experimental: false

Output of docker info:
Containers: 31
 Running: 14
 Paused: 0
 Stopped: 17
Images: 53
Server Version: 17.06.2-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: 2zwq2kcpcycejcl1kyyqfrj1k
 Is Manager: true
 ClusterID: i693q20x89nxzy417a6g6txgx
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Root Rotation In Progress: false
 Node Address: 192.168.0.92
 Manager Addresses:
  192.168.0.92:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: fc48a25bde6fb041aae0977111ad8141ff396438 (expected: 810190ceaa507aa2727d7ae6f4790c76ec150bd2)
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 7.639GiB
```

**Comment by thaJeztah** [comment]
Note that Docker 17.06 (community edition) has reached EOL two years ago, and is
no longer maintained: I highly recommend updating to a current version, because
old versions (including 17.06) have unpatched vulnerabilities (including possible
container escapes; allowing processes in the container to get root access to the
host).


First of all, your healthcheck looks to be incorrect;

```
"Log": [
    {
        "Start": "2019-07-03T15:10:29.3432342Z",
        "End": "2019-07-03T15:10:29.630262Z",
        "ExitCode": 3,
        "Output": "curl: (3) Host name '127.0.0.1 || exit 1' contains bad letter\n"
    },
    {
        "Start": "2019-07-03T15:10:39.6443858Z",
        "End": "2019-07-03T15:10:39.8848617Z",
        "ExitCode": 3,
        "Output": "curl: (3) Host name '127.0.0.1 || exit 1' contains bad letter\n"
    },
    {
        "Start": "2019-07-03T15:10:49.8629749Z",
        "End": "2019-07-03T15:10:50.1144785Z",
        "ExitCode": 3,
        "Output": "curl: (3) Host name '127.0.0.1 || exit 1' contains bad letter\n"
    }
]
```

You can change the healthcheck to just;

```
["CMD", "curl", "-f", "http://127.0.0.1"]
```

Because curl will exit with a non-zero exit code when failing to complete the request

```
$ curl http://bla
curl: (6) Could not resolve host: bla
$ echo $?
6
$ curl http://127.0.0.1:1234
curl: (7) Failed to connect to 127.0.0.1 port 1234: Connection refused
$ echo $?
7
```

I tried reproducing this on a current version of Docker, and was not able to.
Here's the step I took:


First, create an "old" and "new" image (I'm using HEREDOC for convenience):

```bash
docker build -t image-old -<<EOF
FROM nginx:alpine
RUN apk add --no-cache curl
RUN echo old > /usr/share/nginx/html/index.html
EOF


docker build -t image-new -<<EOF
FROM nginx:alpine
RUN apk add --no-cache curl
RUN echo new > /usr/share/nginx/html/index.html
EOF
```

Deploy a stack using that image (slightly simplified from your example. I'm
also using `--resolve-image=never` because the image was not pushed to a registry):

```bash
docker stack deploy --resolve-image=never -c- test <<EOF
version: "3"
services:
  healthcheck:
    image: image-old
    ports:
      -  "2345:80"
    deploy:
      replicas: 2
      update_config:
        parallelism: 1
        delay: 1m
      restart_policy:
        condition: on-failure
    healthcheck:
      test: ["CMD", "curl", "-f", "http://127.0.0.1"]
      interval: 10s
      timeout: 5s
      retries: 3
EOF
```

Wait for the service to be deployed and healthy, then, in another shell, run curl in a loop:

```
while true; do curl 127.0.0.1:2345 ;done

old
old
old
old
old
old
old
old
old
```


Update the service to use the new image:

```bash
docker stack deploy --resolve-image=never -c- test <<EOF
version: "3"
services:
  healthcheck:
    image: image-new
    ports:
      -  "2345:80"
    deploy:
      replicas: 2
      update_config:
        parallelism: 1
        delay: 1m
      restart_policy:
        condition: on-failure
    healthcheck:
      test: ["CMD", "curl", "-f", "http://127.0.0.1"]
      interval: 10s
      timeout: 5s
      retries: 3
EOF
```

See that the services are "rolling" updated; first 1 instance:

```
old
new
old
new
old
new
old
new
old
new
old

```

And after 1 minute both instances:

```
new
new
new
new
new
new
new
new
```


Are you able to reproduce this with the steps above? If so; are you still able
to reproduce on a current version of Docker?

**Comment by arz-GW** [comment]
Thank you very much for your help, but I have modified the yaml file and the problem still exists.


version: "3"
services:
  guwei-healthcheck:
#    image:  uhub.service.ucloud.cn/arz_registry/guwei-healthcheck-new:1
    image: uhub.service.ucloud.cn/arz_registry/guwei-healthcheck-old:1
    ports:
      -  "2345:80"
    volumes:
      - /data/containr/logs/guwei-healthcheck:/var/log/nginx/
    deploy:
      replicas: 2 
      update_config:
        parallelism: 1
        delay: 1m
      restart_policy:
        condition: on-failure
    healthcheck:
      test: ["CMD", "curl", "-f", "http://127.0.0.1"]
      interval: 10s
      timeout: 5s
      retries: 3

while true; do curl 127.0.0.1:2345 ;done

old
old
curl: (56) Recv failure: Connection reset by peer
old
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
old
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
old
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
old
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
old
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
new
new
.....
curl: (7) Failed connect to 127.0.0.1:2345; Connection refused
new
new

**Comment by thaJeztah** [comment]
Is that still on Docker 17.06? There have been various improvements in this area, so if you're still on 17.06, it might be worth trying on a current version

**Comment by arz-GW** [comment]
Yes, that's  still  on Docker 17.06,
I'm planning to use the latest version of docker
Can docker be smoothly updated to the latest version, continuous business?
Thank you

**Comment by thaJeztah** [comment]
The recommendation has always been to upgrade from "stable" to "stable" version (17.06 -> 17.09 -> 17.12 -> 18.03). Bigger steps may work; I'd never do this directly on your production cluster, always test in a staging environment (and have a backup plan in place). Be aware that mixed-version clusters/swarms are not supported (only temporarily during upgrade, but no changes should be made to the cluster state)

If you have a staging or testing environment that also reproduces this problem, I'd try upgrading there to see if the problem is fixed (or setup a temporary environment for testing).

**Comment by arz-GW** [comment]
Thank you. Thank you very much for your help and advice.

I'll try to upgrade the version in the test environment first to see if I can solve the previous problem.

**Comment by arz-GW** [comment]
> The recommendation has always been to upgrade from "stable" to "stable" version (17.06 -> 17.09 -> 17.12 -> 18.03). Bigger steps may work; I'd never do this directly on your production cluster, always test in a staging environment (and have a backup plan in place). Be aware that mixed-version clusters/swarms are not supported (only temporarily during upgrade, but no changes should be made to the cluster state)
> 
> If you have a staging or testing environment that also reproduces this problem, I'd try upgrading there to see if the problem is fixed (or setup a temporary environment for testing).

Hello, I tried docker swarm mode  on version 18.03, the problem is still the same.

new
new
new
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
curl: (7) Failed connect to 192.168.0.98:2345; Connection refused
old
old
old
old
old

docker --version
Docker version 18.03.1-ce, build 9ee9f40


docker info
Containers: 9
 Running: 8
 Paused: 0
 Stopped: 1
Images: 17
Server Version: 18.03.1-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: ascoyz2fw0up8dsc517ci1bji
 Is Manager: false
 Node Address: 192.168.0.90
 Manager Addresses:
  192.168.0.98:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 6
Total Memory: 11.58GiB
Name: 90-docker-worker
ID: EAJU:KAMQ:MVWM:X2K5:4B6U:AIPW:ZAVG:JCOI:SVQP:ICGK:JXLE:BBBN
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 10.19.82.227:5000
 192.168.0.93:5000
 uhub.service.ucloud.cn
 127.0.0.0/8
Live Restore Enabled: false


I wonder if it has something to do with the local TCP connection process, but I changed two test environments and the problem is the same. I wonder how the services in docker swarm can be smoothly updated and user access is imperceptible.
